<!DOCTYPE html>
<html lang="en"><head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1"><!-- Begin Jekyll SEO tag v2.8.0 -->
<title>Eremid Genomics and Catawba College – Illumina Next Generation Sequencing Workshop | PWS Notes</title>
<meta name="generator" content="Jekyll v3.9.5" />
<meta property="og:title" content="Eremid Genomics and Catawba College – Illumina Next Generation Sequencing Workshop" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="Useful Links" />
<meta property="og:description" content="Useful Links" />
<link rel="canonical" href="http://localhost:4000/2024/08/12/eremid-nextgen-workshop.html" />
<meta property="og:url" content="http://localhost:4000/2024/08/12/eremid-nextgen-workshop.html" />
<meta property="og:site_name" content="PWS Notes" />
<meta property="og:type" content="article" />
<meta property="article:published_time" content="2024-08-12T00:00:00-04:00" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="Eremid Genomics and Catawba College – Illumina Next Generation Sequencing Workshop" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"BlogPosting","dateModified":"2024-08-12T00:00:00-04:00","datePublished":"2024-08-12T00:00:00-04:00","description":"Useful Links","headline":"Eremid Genomics and Catawba College – Illumina Next Generation Sequencing Workshop","mainEntityOfPage":{"@type":"WebPage","@id":"http://localhost:4000/2024/08/12/eremid-nextgen-workshop.html"},"url":"http://localhost:4000/2024/08/12/eremid-nextgen-workshop.html"}</script>
<!-- End Jekyll SEO tag -->
<link rel="stylesheet" href="/assets/main.css"><link type="application/atom+xml" rel="alternate" href="http://localhost:4000/feed.xml" title="PWS Notes" /></head>
<body><header class="site-header" role="banner">

  <div class="wrapper"><a class="site-title" rel="author" href="/">PWS Notes</a><nav class="site-nav">
        <input type="checkbox" id="nav-trigger" class="nav-trigger" />
        <label for="nav-trigger">
          <span class="menu-icon">
            <svg viewBox="0 0 18 15" width="18px" height="15px">
              <path d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.032C17.335,0,18,0.665,18,1.484L18,1.484z M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.032C17.335,6.031,18,6.696,18,7.516L18,7.516z M18,13.516C18,14.335,17.335,15,16.516,15H1.484 C0.665,15,0,14.335,0,13.516l0,0c0-0.82,0.665-1.483,1.484-1.483h15.032C17.335,12.031,18,12.695,18,13.516L18,13.516z"/>
            </svg>
          </span>
        </label>

        <div class="trigger"><a class="page-link" href="/about/">About</a><a class="page-link" href="/blog.html">Blog</a></div>
      </nav></div>
</header>
<main class="page-content" aria-label="Content">
      <div class="wrapper">
        <article class="post h-entry" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title p-name" itemprop="name headline">Eremid Genomics and Catawba College -- Illumina Next Generation Sequencing Workshop</h1>
    <p class="post-meta">
      <time class="dt-published" datetime="2024-08-12T00:00:00-04:00" itemprop="datePublished">Aug 12, 2024
      </time></p>
  </header>

  <div class="post-content e-content" itemprop="articleBody">
    <h2 id="useful-links">Useful Links</h2>

<ol>
  <li><a href="https://www.geneious.com/free-trial">Geneious Prime Free Trial</a></li>
  <li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a></li>
  <li><a href="https://cornell.box.com/s/o5y164l09syhttirifhv3ow1qhsht1en">Sequence Data Files (Box Drive)</a></li>
  <li><a href="https://learn.microsoft.com/en-us/windows/wsl/">Windows Subsystem for Linux</a></li>
  <li><a href="https://chlorobox.mpimp-golm.mpg.de/geseq.html">GeSeq</a></li>
  <li><a href="https://www.bv-brc.org/app/TaxonomicClassification">BV-BRC Taxonomic Classification Service</a></li>
  <li><a href="https://www.phylo.org">CIPRES Science Gateway</a></li>
  <li><a href="https://usegalaxy.org/">Galaxy Webserver</a></li>
  <li><a href="https://docs.anaconda.com/miniconda/#quick-command-line-install">Miniconda Install</a></li>
</ol>

<h2 id="unixlinux-command-line-plastome-assembly">Unix/Linux Command Line Plastome Assembly</h2>

<h3 id="software">Software</h3>

<ul>
  <li><a href="https://github.com/OpenGene/fastp">fastp</a></li>
  <li><a href="https://github.com/ablab/spades">SPAdes</a></li>
  <li><a href="https://github.com/Kinggerm/GetOrganelle">getOrganelle</a></li>
  <li><a href="https://github.com/ndierckx/NOVOPlasty">NOVOPlasty</a></li>
  <li><a href="https://github.com/iqtree/iqtree2">iqtree</a></li>
</ul>

<h3 id="install">Install</h3>

<p>Install conda if you haven’t already. Following are instructions for Linux and Mac – pick correct one for your computer. Follow the on-screen prompts using default settings.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># MacOS ARM CPU
mkdir -p ~/miniconda3
curl https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-arm64.sh -o ~/miniconda3/miniconda.sh
bash ~/miniconda3/miniconda.sh -b -u -p ~/miniconda3
rm -rf ~/miniconda3/miniconda.sh

# MacOS Intel CPU
curl https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh -o ~/miniconda3/miniconda.sh

# Linux
mkdir -p ~/miniconda3
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O ~/miniconda3/miniconda.sh
bash ~/miniconda3/miniconda.sh -b -u -p ~/miniconda3
rm -rf ~/miniconda3/miniconda.sh
</code></pre></div></div>

<p>Intialize</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>~/miniconda3/bin/conda init bash
~/miniconda3/bin/conda init zsh
</code></pre></div></div>

<p>Restart your terminal session for it to take effect.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>conda install -c bioconda -c conda-forge fastp spades getorganelle novoplasty iqtree matplotlib
</code></pre></div></div>

<h3 id="inspect-data">Inspect Data</h3>

<p>Trim adapters, discard low quality reads with fastp.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>fastp -i R1.fastq.gz -I R2.fastq.gz -o R1.fastp.fastq.gz -O R2.fastp.fastq.gz -5 -3 --detect_adapter_for_pe -p --html SAMPLE.html --json SAMPLE.json
</code></pre></div></div>

<p>Inspect fastp html file, check for any abnormalities.</p>

<p>(Optional) Taxonomic classification to check for bacterial contaminants. https://www.bv-brc.org/app/TaxonomicClassification</p>

<h3 id="method-1-reference-based-filtering-followed-by-de-novo-assembly">Method 1: Reference-based filtering followed by de novo assembly</h3>

<p>Get a complete reference sequence (FASTA format) from NCBI GenBank or other source.</p>

<p>Index reference sequence with Bowtie2.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bowtie2-build reference.fasta reference.fasta
</code></pre></div></div>

<p>Map reads to reference, saving mapped reads to new files. Adjust <code class="language-plaintext highlighter-rouge">-p</code> to match your computer’s # CPUs. Using <code class="language-plaintext highlighter-rouge">--very-sensitive-local</code> enhances short matches for more distantly related species.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bowtie2 -p 12 -x reference.fasta -1 R1.fastp.fastq.gz -2 R2.fastp.fastq.gz --al-conc-gz R%.aln.fastq.gz --very-sensitive-local &amp;&gt; /dev/null
</code></pre></div></div>

<p>Assemble mapped reads with SPAdes. Change <code class="language-plaintext highlighter-rouge">-t</code> to match your computers # of CPUs. Change <code class="language-plaintext highlighter-rouge">-o</code> to whatever the sample name is.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>spades.py -1 R1.aln.fastq.gz -2 R2.aln.fastq.gz -t 12 -o SAMPLE_NAME --careful
</code></pre></div></div>

<p>Check the <code class="language-plaintext highlighter-rouge">scaffolds.fasta</code> file in the output folder. If successful, there should be three large scaffolds, approx. 100k, 25k, and 15k in length, representing the LSC, SSC, and IR regions. Manually stitch them together in Geneious based on the reference. 
If more and/or smaller scaffolds are produced, try mapping them to the reference in Geneious to see what’s missing.</p>

<h3 id="method-2-completely-de-novo-with-getorganelle">Method 2: Completely de novo with getOrganelle</h3>

<p>Input all adapter-trimmed reads into getOrganelle. To set up getOrganelle for the first time:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># This downloads all databases for different organisms and organelles. See help menu/documentation to download certain ones.
get_organelle_config.py -a all
</code></pre></div></div>
<p>Run getOrganelle for an embryophyte plant plastid. See documention for other organism and organelle options.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>get_organelle_from_reads.py -1 R1.fastp.fastq.gz -2 R2.fastp.fastq.gz -F embplant_pt -t 12 -o output_dir
</code></pre></div></div>

<p>Inspect output files. Hopefully you’ll get a complete circular sequence with filename like: <code class="language-plaintext highlighter-rouge">embplant_pt.K115.complete.graph1.1.path_sequence.fasta</code>.</p>

<p>Remap reads to assembled plastome to evaluate.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>evaluate_assembly_using_mapping.py -f embplant_pt.K115.complete.graph1.1.path_sequence.fasta -1 R1.fastp.fastq.gz -2 R2.fastp.fastq.gz -t 12 -o evaluate_output --draw
</code></pre></div></div>

<p>Check coverage levels plotted in <code class="language-plaintext highlighter-rouge">mapping.pdf</code>. “Matched”  level should be pretty even (though noisy) across LSC/SSC, maybe be different in IRs.</p>

<h3 id="method-3-completely-de-novo-with-novoplasty">Method 3: Completely de novo with NOVOPlasty.</h3>

<p>NOVOPlasty requires a starting sequence (FASTA format) that closely matches your organism. Find a sequence for a conserved gene (e.g. rbcL) from GenBank or another source. Copy the file into your working directory.</p>

<p>//TODO</p>

  </div><a class="u-url" href="/2024/08/12/eremid-nextgen-workshop.html" hidden></a>
</article>

      </div>
    </main><footer class="site-footer h-card">
  <data class="u-url" href="/"></data>

  <div class="wrapper">

    <h2 class="footer-heading">PWS Notes</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li class="p-name">PWS Notes</li><li><a class="u-email" href="mailto:peter@peterwschafran.com">peter@peterwschafran.com</a></li></ul>
      </div>

      <div class="footer-col footer-col-2"><ul class="social-media-list"><li><a href="https://www.facebook.com/peter.schafran"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#facebook"></use></svg> <span class="username">peter.schafran</span></a></li><li><a href="https://github.com/pschafran"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#github"></use></svg> <span class="username">pschafran</span></a></li><li><a href="https://instagram.com/peter_schafran"><svg class="svg-icon"><use xlink:href="/assets/minima-social-icons.svg#instagram"></use></svg> <span class="username">peter_schafran</span></a></li></ul>
</div>

      <div class="footer-col footer-col-3">
        <p>I am an evolutionary biologist and systematist interested in non-model non-flowering plants, especially the order Isoetales (quillworts) and phylum Anthocerotophyta (hornworts). This blog documents notes and code that don’t fit into standard publications.</p>
      </div>
    </div>

  </div>

</footer>
</body>

</html>
